nkms/policy/models.py

```diff
@@ -0,0 +1,107 @@
+from nkms.crypto.hash import content_hash
+from nkms.crypto.pre.keygen import generate_re_encryption_keys
+from nkms.policy.constants import UNKNOWN_KFRAG
+
+
+class PolicyGroup(object):
+    """
+    The terms and conditions by which Alice shares with Bob.
+    """
+
+    def __init__(self, policies=None):
+        self.policies = policies or []
+
+    @staticmethod
+    def craft(keychain_alice: "KeyChain",
+              pubkey_enc_bob: tuple,
+              uri: bytes,
+              m: int,
+              n: int
+              ):
+        """
+        Alice dictates a new group of policies.
+        """
+        re_enc_keys = generate_re_encryption_keys(keychain_alice.enc_keypair.priv_key,
+                                                  pubkey_enc_bob,
+                                                  m,
+                                                  n)
+        policies = []
+        for kfrag_id, key in enumerate(re_enc_keys):
+            policy = Policy.from_alice(
+                key,  # Bob won't know this.
+                keychain_alice.sig_keypair.pub_key,
+                pubkey_enc_bob,
+                uri,  # Ursula won't know this.
+                kfrag_id,
+            )
+            policies.append(policy)
+
+        return PolicyGroup(policies)
+
+
+    def transmit(self, networky_stuff):
+        for policy in self.policies:
+            policy_offer = policy.craft_offer(networky_stuff)
+            result = networky_stuff.transmit_offer(policy.ursula, policy_offer)
+            if result.was_accepted:
+                policy.update_treasure_map(result)
+
+
+class Policy(object):
+    """
+    An individual agreement between Alice and Ursula.  Together, all of the Policies by which
+    Ursula nodes which enter into an agreement regarding the same series of kFrags constitute
+    a PolicyGroup.
+
+    A Policy has a unique ID, which includes a fingerprint of Alice's public key so that
+    only she can set a policy with that ID.  Ursula must verify this; otherwise a collision
+    attack is possible.
+    """
+    ursula = None
+    hashed_part = None
+
+    def __init__(self, kfrag=UNKNOWN_KFRAG, challenge_size=20):
+        self.kfrag = kfrag
+        self.challenge_size = challenge_size
+        self.treasure_map = []
+
+    @staticmethod
+    def from_alice(kfrag,
+                   pubkey_sig_alice,
+                   pubkey_enc_bob,
+                   uri,
+                   kfrag_id
+                   ):
+        policy = Policy(kfrag)
+        policy.generate_challenge_pack()
+        policy.hash(pubkey_sig_alice=pubkey_sig_alice, hash_input=(pubkey_enc_bob, uri, kfrag_id))
+
+        return policy
+
+    def hash(self, pubkey_sig_alice, hash_input):
+        hash_input = str(hash_input).encode()
+        self.hashed_part = content_hash(hash_input)
+        hash_input_for_id = str(pubkey_sig_alice).encode() + str(self.hashed_part).encode()
```

Comment:
Instead of encoding this, you can hash it and, instead of returning the raw digest, you can call `hexdigest` to get the stringified version of it in hexidecimal.

Reply:
```>>> sha3.keccak_256(b'this is a test').hexdigest()
'9fd09c38c2a5ae0a0bcd617872b735e37909ccc05c956460be7d3d03d881a0dc'
```

Reply:
Oh yeah, I agree that this is a bit of a shitshow.  :-)

But I'm not clear on what you're suggesting here.  Your sample shows a bytes object being passed to `keccak_256()`, but how to proceed for an arbitrary object?  For example, in this case, `hash_input` will often be a tuple.  And then, on line 84, how to properly do this concatenation?

Reply:
The `keccak_256` function will always need a bytes object to hash it, so I recommend taking each element of `hash_input`, encoding it, then hash it to get a bytes object for that element. Do this for each element, then hash them all together and get a `hexdigest` version of it to avoid weird nuances in the string when converting them.

For example:
```
keccak(keccak(pubkey_enc_bob).digest() + keccak(uri).digest() + keccak(kfrag_id).digest).hexdigest()
```

Reply:
This allows two things:
1) You avoid the string conversion which can add weird elements to the hashed string.
2) Allows the hashed results to be more replicatable.

Reply:
Personally, I would have the hashing functions accept a list of things to hash. Check if the type is of `bytes`. If it is not, encode it. Then update the hash via `update(..)`.

I would also include a boolean param like `hex_format` which will cause the hash output to be `hexdigest()` instead of `digest()`.

Reply:
Btw should the hash_input_for_id be necessarily string (hexdigest) not bytes (digest)?